---
title: "Format UCSC cellbrowser"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Format seurat objects into a UCSC cellbrowser. Note requires UCSC (`cellbrowser python package`[ https://cellbrowser.readthedocs.io/index.html]) 

```{r}
source(here::here("R", "utils.R"))
library(scbp)
library(tidyverse)
library(qs)
library(Seurat)
library(glue)
cb_outdir <- file.path("cellbrowser")
dir.create(cb_outdir, showWarnings = FALSE, recursive = TRUE)
def_embeddings <- function(so) {str_subset(names(so@reductions), "umap")}

write_config <- function(x, path){
  glue('{key}="{value}"', 
     key = names(x), 
     value = x) %>% 
    write_lines(path)
}

qread <- function(file, ...) {qs::qread(file, ..., nthreads = 4, use_alt_rep = FALSE)}
```
# Collection overview

```{r}
collection_cb <- c(shortLabel="single cell analysis of club cell regeneration")


collection_desc <- c(title="single cell analysis of club cell regeneration",
                     abstract="To identify genes and signaling pathways regulating the dysfunction of club cells in COPD, we have recently established an inducible double labeling mouse model (Scgb1a1-CreERT;Rosa-dtTomato;TCF-GFP), which allows lineage tracing of club cells by dtTomato expression as well as real-time labeling of the Wnt/β-catenin activity with EGFP expression in these cells. Here, we treat this mouse line with elastase and saline to generate COPD and control mice, respectively, followed by isolation of tdTomato labeled club cells for single cell RNA sequencing analysis. We captured 2 samples (1 sample per treatment, 2000 cells pooled from 3 animals per sample) using the 10X Genomics 3’ Platform. tdtomato and egfp expression can be queried using the names tdtomato-sv40 and egfp-sv40 respectively. The cluster labels, when clicked, will show the top markers of each subpopulation",
                     unitDesc="Log-normalized counts")

write_config(collection_cb, file.path(cb_outdir, "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "desc.conf"))
```


## All cells


```{r}
so <- qread(file.path("objects", "so.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  Treatment = "treatment",
  `Coarse cell type` = "coarse_cell_type",
  `Refined cell type` = "cell_type_cluster",
  `Cell cycle phase` = "Phase")

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "Treatment",
                    "Cell cycle phase"), 
                 project = "All-cells",
                 outdir = cb_outdir,
                 marker_file = "markers/cell_types_refined_cluster_markers.tsv",
                 ident = "Refined cell type",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 config = list(priority = 1,  
                               radius = 4),
                 description = list(    
                   title = "Club cells regeneration",
                   description = "tdtomato and egfp expression can be queried using the names tdtomato-sv40 and egfp-sv40 respectively. The cluster labels, when clicked, will show the top markers of each subpopulation."
                 )
)
```

# Build all 

```{r, eval = FALSE}
cb_conf <- read_lines("~/.cellbrowser.conf") %>% 
  str_subset("dataRoot") %>% 
  str_split("=") %>% 
  .[[1]] %>% 
  .[2] %>% 
  gsub('\"', "", .)

if(cb_outdir != cb_conf ) {
  write_lines(paste0('dataRoot="', 
                     normalizePath(cb_outdir), '"'),
              "~/.cellbrowser.conf")
}
```

```{r}

datasets <- c(
   "All-cells"
)

datasets <- file.path(cb_outdir, datasets, "cellbrowser.conf")

build_cellbrowser(datasets, 
                    outdir = file.path(cb_outdir, "club-regen"),
                    cbBuild_path =  "/miniconda3/envs/py37/bin/cbBuild")
```




